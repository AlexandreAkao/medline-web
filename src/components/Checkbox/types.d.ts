type InputProps = import('react').InputHTMLAttributes<HTMLInputElement>;

interface ICheckboxProps extends InputProps {
  children?: React.ReactNode;
  primary?: boolean;
}
