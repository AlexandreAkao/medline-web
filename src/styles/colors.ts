const colors = {
  primary: {
    normal: '#248F8D',
    light: '#279d9a',
    lighter: '#E9F4F4',
    darker: '#234342',
    darkest: '#257472',
  },
  white: {
    normal: '#fff',
    light: '#ccc',
    lighter: '#f0f0f0',
    lightest: '#fcfcfc',
  },
  black: {
    normal: '#000',
    soft: '#1d2129',
  },
  grey: {
    normal: '#808080',
    dark: '#3d3d3d',
  },
  blue: {
    normal: '#5041FF',
    light: '#0075FF',
  },
};

export default colors;
