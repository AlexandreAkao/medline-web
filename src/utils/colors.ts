export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
