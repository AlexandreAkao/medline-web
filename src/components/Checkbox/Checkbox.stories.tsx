import { ComponentStory, ComponentMeta } from '@storybook/react';

import Checkbox from 'components/Checkbox';
import colors from 'styles/colors';

export default {
  title: 'Checkbox',
  args: {},
  component: Checkbox,
  decorators: [
    Story => (
      <div
        style={{
          background: colors.primary.normal,
          width: '100%',
          height: '100%',
          padding: 20,
        }}
      >
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = args => <Checkbox {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Primary',
  id: 'primary',
  primary: true,
  onChange: event => console.log(event.target.checked),
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: 'Secondary',
  id: 'secondary',
  primary: false,
};
