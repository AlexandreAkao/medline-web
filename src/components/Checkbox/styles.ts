import styled from 'styled-components';

import colors from 'styles/colors';

export const CheckboxInput = styled.input`
  height: 20px;
  width: 20px;
  cursor: pointer;
`;

export const CheckboxLabel = styled.label`
  margin-left: 8px;
  font-weight: 600;
  font-size: 1.125rem;
  cursor: pointer;
`;

export const CheckboxContainer = styled.div<ICheckboxProps>`
  display: inline-flex;
  justify-content: center;
  align-items: center;
  height: 20px;

  ${CheckboxLabel} {
    color: ${({ primary }) => (primary ? colors.black.soft : colors.white.normal)};
  }
`;
