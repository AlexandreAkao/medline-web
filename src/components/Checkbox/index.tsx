import { CheckboxContainer, CheckboxInput, CheckboxLabel } from 'components/Checkbox/styles';

function Checkbox({ id, children, primary, ...rest }: ICheckboxProps) {
  return (
    <CheckboxContainer primary={primary}>
      <CheckboxInput {...rest} type="checkbox" name="" id={id} />
      <CheckboxLabel htmlFor={id}>{children}</CheckboxLabel>
    </CheckboxContainer>
  );
}

export default Checkbox;
