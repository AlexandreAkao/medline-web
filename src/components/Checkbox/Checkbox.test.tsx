import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import * as stories from 'components/Checkbox/Checkbox.stories';
import Checkbox from 'components/Checkbox';
import colors from 'styles/colors';

const { Primary, Secondary } = composeStories(stories);

describe('Checkbox', () => {
  const renderComponent = (customProps: Partial<ICheckboxProps> = {}) => {
    const props: ICheckboxProps = {
      children: 'checkbox-test',
      id: 'checkbox-test-id',
      ...customProps,
    };
    return render(<Checkbox {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a primary checkbox', () => {
    render(<Primary />);
    const labelElement = screen.getByText(/Primary/i);
    expect(labelElement).toBeInTheDocument();
    expect(labelElement).toHaveStyle({
      color: colors.black.soft,
    });
  });

  it('should render a secondary checkbox', () => {
    render(<Secondary />);
    const labelElement = screen.getByText(/Secondary/i);
    expect(labelElement).toBeInTheDocument();
    expect(labelElement).toHaveStyle({
      color: colors.white.normal,
    });
  });

  it('should trigger function on change checkbox value', async () => {
    const onChange = vi.fn();
    renderComponent({ onChange });

    const checkboxElement = screen.getByRole('checkbox');
    const labelElement = screen.getByText(/checkbox-test/i);

    expect(labelElement).toBeInTheDocument();
    expect(checkboxElement).toBeInTheDocument();
    expect(checkboxElement).not.toBeChecked();

    await userEvent.click(checkboxElement);
    expect(checkboxElement).toBeChecked();
    expect(onChange).toBeCalled();

    await userEvent.click(labelElement);
    expect(checkboxElement).not.toBeChecked();
    expect(onChange).toBeCalled();
  });
});
