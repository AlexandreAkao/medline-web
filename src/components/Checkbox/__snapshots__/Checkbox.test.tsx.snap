// Vitest Snapshot v1

exports[`Checkbox > should match with snapshot 1`] = `
<div>
  <div
    class="sc-dkzDqf eMxdaI"
  >
    <input
      class="sc-bczRLJ itSYxX"
      id="checkbox-test-id"
      name=""
      type="checkbox"
    />
    <label
      class="sc-gsnTZi cTEWgn"
      for="checkbox-test-id"
    >
      checkbox-test
    </label>
  </div>
</div>
`;
